Make planner node an action server

p-nayak11 opened the pull request:
## Description
Turns planner node into an Action Server.

## Associated / Resolved Issue
 [ClickUp card](https://app.clickup.com/t/2zx11wp)

## Design Documents
[Link](https://docs.google.com/document/d/1qnf3ca5nMZWkTeyERrk6fsng_ZnlourK01f8Z9vLoPs/edit?usp=sharing)

## Steps to Test
### Test Case 1
1. Run sim.
2. Observe Robot 2 heading to point and heading [given by the ActionClient](https://github.com/RoboJackets/robocup-software/pull/1948/files#diff-6ca09102ecda0a74b48b3688563a3f61800650f3cd6c962190122cb83fe9d12eR38-R45) (client will be rewritten in full by @kfu02 )
3. Ensure console outputs is_done when done.

**Expected result:** robot 2 avoids obstacles when going to its point (but only the first time).

## Key Files to Review
Action Server / Client
 * planner_node.*pp
 * robot_action_client.*pp

## Review Checklist

- [x] **Docstrings**: All methods and classes should have the file appropriate docstrings which follow the guidelines in the ["Contributing" page](https://rj-rc-software.readthedocs.io/en/latest/contributing.html) of our docs.
- [x] **Remove extra print statements**: Any print statements used for debugging should be removed
- [x] **Tag reviewers**: Tag some people for review and ping them on Slack

p-nayak11 committed:
Added necessary msg and action files

Commit:
scaffolded planner node -> action server

Commit:
WIP add action client

Commit:
WIP add action client

kfu02 commented on rj_msgs/msg/IsDone.msg:
```diff
@@ -0,0 +1 @@
+bool is_done
```
@p-nayak11 I realized you're right, this msg isn't necessary anymore. we can delete it next time

fourpenny committed:
WIP action client

p-nayak11 committed:
add hpp files

p-nayak11 committed:
Merge branch 'ros2' into planner_action_server

p-nayak11 committed:
working action client!

p-nayak11 committed:
working action client!

p-nayak11 committed:
restore two team control (#1937)

* make network radio port respond to ROS param

* rm unused radio_node.cpp

* make soccer.launch.py take in separate yamls for network/other params

- separates sim/real networking param files from other params
- adds new alt param for multiple broadcast server_ports

* TODOs and make pretty

* started real2play

* replace broken real2play launch file with bash script

* rm excessive extra .yamls, replace with server_port param
 - also sets make real2play = launch bash script

* cleanup

* rm kNetworkRadioServerPort

* replace sim2play.launch with bash, like real2play

* Revert "replace sim2play.launch with bash, like real2play"

This reverts commit 8743c8dbf1c517cbfaeb649a86bffb43be5a888d.

* mv param to sim radio where it is used

* use param_config_yaml everywhere

* do param config less janky, revive bash script sim2play

* add comments to launch

* add some more comments

* comply with GH Actions style flags

* clarify comment

* add team name as launch flag, make run_sim flag control yaml used

* bash script doesn't really work, so add domain id to makefile

* fix launching sim

* make pretty-lines

Co-authored-by: RoboCup Field Comp <fake-field-comp-email@not-real.com>
Co-authored-by: kasohrab <kasra.sohrab@gmail.com>

p-nayak11 committed:
finish rebasing

p-nayak11 committed:
Merge branch 'planner_action_server' of https://github.com/RoboJackets/robocup-software into planner_action_server

p-nayak11 committed:
Motion planning commands executed through client-server model

p-nayak11 committed:
add feedback & make pretty lines

kfu02 committed:
add is_done, fix execute loop implementation, add TODOs

kfu02 commented on soccer/src/soccer/planning/planner/path_target_planner.cpp:
```diff
@@ -62,7 +62,7 @@ bool PathTargetPlanner::is_done() const {
     // consider close enough (in m)
     // TODO(#1913): connect gameplay to planner is_done to avoid two diff threshold params
-    double temp_correction = 1.2;  // be X% more generous than gameplay so we can see change
+    double temp_correction = 1.0;  // be X% more generous than gameplay so we can see change
```
I will fix these when writing the new action clients.

kfu02 commented on soccer/src/soccer/planning/planner_node.cpp:
```diff
@@ -27,6 +44,76 @@ PlannerNode::PlannerNode()
+                                                     std::shared_ptr<const RobotMove::Goal> goal) {
+    (void)uuid;
+
+    // TODO(p-nayak): REJECT duplicate goal requests so we aren't constantly replanning them
```
@p-nayak11 can you make a card for your TODOs?

kfu02 commented on soccer/src/soccer/robot_action_client/robot_action_client.cpp:
```diff
@@ -0,0 +1,95 @@
+#include "robot_action_client.hpp"
```
I will rewrite this.

kfu02 commented on soccer/src/soccer/robot_action_client/robot_action_client_main.cpp:
```diff
@@ -0,0 +1,12 @@
+    rclcpp::init(argc, argv);
+    rj_utils::set_spdlog_default_ros2("processor");
+
+    SPDLOG_ERROR("Robot action client is being spun");
```
```suggestion
```

kfu02 reviewed (commented):
@p-nayak11 and I paired for most of this, and I did some work alone at the end, so I can't leave a real review on this. Here's some comments to help other reviewers.

kfu02 committed:
rm debug print

kasohrab commented on soccer/src/soccer/planning/planner_node.cpp:
```diff
@@ -27,6 +44,76 @@ PlannerNode::PlannerNode()
+
+void PlannerNode::execute(const std::shared_ptr<GoalHandleRobotMove> goal_handle) {
+    // rate-limit loop to 1ms per iteration
+    rclcpp::Rate loop_rate(1);
```
```suggestion
    rclcpp::Rate loop_rate(1ms);
```
to actually use ms

kasohrab commented on soccer/src/soccer/planning/planner_node.cpp:
```diff
@@ -27,6 +44,76 @@ PlannerNode::PlannerNode()
+                goal_handle->succeed(result);
+                return;
+            }
+        }
```
```suggestion
        }
        loop_rate.sleep();
```

kasohrab replied:
It would be nice to include this as part of our RJ::Time eventually.

kasohrab replied:
this change is so that we actually use the loop rate which is important for performance (especially on a laptop with 6 robots).

kasohrab commented on soccer/src/soccer/planning/planner_node.cpp:
```diff
@@ -27,6 +44,76 @@ PlannerNode::PlannerNode()
+    rclcpp::Rate loop_rate(1);
+
+    // create ptrs to Goal, Result objects per ActionServer API
+    std::shared_ptr<const RobotMove::Goal> goal = goal_handle->get_goal();
```
I'm not sure why it needs to be a shared ptr here, but I'm happy to ignore it for now.

kasohrab commented on soccer/src/soccer/planning/planner_node.cpp:
```diff
@@ -27,6 +44,76 @@ PlannerNode::PlannerNode()
+    std::shared_ptr<PlannerForRobot> my_robot_planner = robots_planners_[robot_id];
+
+    // loop until goal is done (SUCCEEDED or CANCELED)
+    bool not_done_yet = true;  // yes, this is necessary (compiler optimizes out while(true))
+    while (not_done_yet) {
```
```suggestion
    for (;;) {
```
It worked for me like this. I think it is more clear this way that the loop will only end on external factors.

kfu02 replied:
oh sweet, good to know

kasohrab reviewed (changes requested):
looks great! just some small requests before merging

kfu02 committed:
Apply suggestions from code review

Co-authored-by: Kasra Alexander Sohrab <67869617+kasohrab@users.noreply.github.com>

kfu02 commented:
merging despite some check warnings from GH Actions. ROS2 action server spec prefers binds to lambdas, and specifies whether our params are const or not.
